selfdrive/test/process_replay/process_replay.py

```diff
@@ -353,3 +380,39 @@ def replay_process(cfg, lr):
 
         recv_cnt -= m.which() in recv_socks
   return log_msgs
+
+def cpp_replay_process(config, logreader):
+  pub_sockets = [s for s in config.pub_sub.keys()]  # We dump data from logs here
+  sub_sockets = [s for _, sub in config.pub_sub.items() for s in sub]  # We get responses here
+  pm = messaging.PubMaster(pub_sockets)
+  sockets = {s : messaging.sub_sock(s) for s in sub_sockets}
+  print("Sorting logs")
+  all_msgs = sorted(logreader, key=lambda msg: msg.logMonoTime)
+  pub_msgs = [msg for msg in all_msgs if msg.which() in list(config.pub_sub.keys())]
+
+  manager.start_managed_process(config.proc_name)
+
+  time.sleep(5)  # We give the process time to start
+
+  log_msgs = []
+  for s in sub_sockets:
+    messaging.recv_one_or_none(sockets[s])
+
+  time.sleep(1)
+  for msg in tqdm(pub_msgs):
+    pm.send(msg.which(), msg.as_builder())
+    for s in config.should_recv_callback(msg):
+      ctime = time.time()
+      response = None
+      while response is None and time.time() - ctime < 2.5:
```

Comment:
use timeout argument on socket creation instead of manually implementing this. This uses 100% of a core while waiting.